### DISCUSSION TITLE
Workflow_run doen't preserve github.ref?

### DISCUSSION THREAD

**Discussion by rducom** [community/community#27124] [Actions]
<p>Hello,</p>
<p>I have 2 workflows linked by the workflow_run feature.<br>
When creating a release (= when pushing a git tag) :</p>
<ul>
<li>on the main workflow my github.ref is <code>"ref": "refs/tags/v0.2.0"</code>
</li>
<li>but on the “workflow_run” one, my github.ref is still <code>"ref": "refs/heads/master"</code>…</li>
</ul>
<p>Both ref are on the same sha.</p>
<p>My goal is to keep the “release” part of my workflows isolated, and trigger the “workflow_run” one only when tags are pushed AND the primary workflow completes successfully.</p>
<p>first one :</p>
<pre><code class="lang-auto">name: .NET Core
on:
  push:
    tags:
      - '**'
    branches:
      - master
  pull_request:
</code></pre>
<p>second one :</p>
<pre><code class="lang-auto">name: Upload Release Asset
on:
  workflow_run:
    workflows: [".NET Core"]
    types:
      - completed
</code></pre>
<p>Is this a known bug ? a known limitation ? is there a workaround ?</p>

**Comment by rducom** [comment]
<p>I created a simple repro :</p>
<ul>
<li>main workflow with correct git ref : <a href="https://github.com/rducom/repro_bug_Workflow_run/runs/1695633669?check_suite_focus=true#step:4:198" class="inline-onebox" rel="noopener nofollow ugc">fix default branch · rducom/repro_bug_Workflow_run@135a6b8 · GitHub</a>
</li>
<li>dependent workflow with incorrect git ref : <a href="https://github.com/rducom/repro_bug_Workflow_run/runs/1695635532?check_suite_focus=true#step:4:350" class="inline-onebox" rel="noopener nofollow ugc">fix default branch · rducom/repro_bug_Workflow_run@135a6b8 · GitHub</a>
</li>
</ul>

**Comment by otter-computer** [comment]
<p>I think this is just a limitation of how this feature works with respect to the way you’re using it. The second workflow is completely independent of the first and will receive a brand new webhook event to trigger it meaning it will just be triggered using the “default” git ref.</p>
<p>Is there a specific reason you’re wanting keep this in a completely separate file? For this situation I’d probably recommend just using a separate job inside the same workflow file. You can use the <code>needs</code> key to prevent your release job from running until your first job completes, and this should preserve the git ref too as it will still be the same webhook event that triggered the workflow.</p>

**Comment by rducom** [comment]
<blockquote>
<p>The second workflow is completely independent of the first</p>
</blockquote>
<p>I disagree, the second workflow depends of the first, that’s the intention of the "workflow_run’ feature: let workflows depends on others. Or I completely miss the point.</p>
<p>In my specific case, I need to know there’s a git tag. (More precisely, I want to know the workflow is running because of a tag.). So the git ref is mandatory.</p>
<p>The specific reason I would like to have two separates workflow is just CI structure : having smaller workflow files, and let run specific workflows (like a “publish/release artifact” run only when needed).<br>
I know I can “just put everything in one workflow”, but it seems really more convenient to have distinct workflow, and just run what’s needed when needed.</p>
<p>Moreover, it would allow us to re-use previous parts of the the build. Ex : if I have 5 workflows, and each one need to build and test the whole repository before doing the “remaining part of the job”, it looks like a major resource waste as I could have just build and test the commit once, and then run the “other things” in the 4 or 5 dependent workflows…</p>
<p>That’s why I’m convinced git ref should be propagated from “main” workflows to “dependent workflows”.</p>

**Comment by rducom** [comment, marked as answer]
<p>Thanks to the GH support, the answer is here : <a href="https://github.community/t/workflow-run-not-working-as-expected/139342/4" class="inline-onebox">Workflow_run not working as expected - #4 by brightran</a></p>
<p>We can access the original ref with</p>
<blockquote>
<p>github.event.workflow_run.head_branch</p>
</blockquote>
<p>Problem solved ! :rocket:</p>

**Comment by otter-computer** [comment]
<aside class="quote no-group" data-username="rducom" data-post="4" data-topic="155148">
<div class="title">
<div class="quote-controls"></div>
 rducom:</div>
<blockquote>
<p>I disagree, the second workflow depends of the first, that’s the intention of the "workflow_run’ feature: let workflows depends on others. Or I completely miss the point.</p>
</blockquote>
</aside>
<p>What I meant by this was that as this was an entirely separate workflow it would be treated as such. The variables set in this run would be set at start time and not be inherited from the original.</p>
<p>But it did slip my mind that you’d be able to check inside the triggering webhook event, so I’m glad you managed to find the answer!</p>
